### PR TITLE
Refuse a date the calendar never reaches

### PR DESCRIPTION
**Pull request by bartblast**
Closes #1184

## Summary by CodeRabbit

* **Bug Fixes**
  * Date, datetime, and time values are now validated against actual calendar rules.
  * Invalid months, days, leap dates, times, microseconds, and precisions are rejected consistently.
  * Client-side date and datetime parsing now refuses malformed or impossible values.
  * Valid boundary values, including leap-year dates and maximum times, continue to be accepted.

* **Tests**
  * Added coverage for invalid temporal values and browser-side rejection without submitting invalid data.

### CODE REVIEW

**Comment by bartblast** [comment]
@coderabbitai pause

File: assets/js/elixir/hologram/entity.mjs [hunk 1]
```diff
@@ -315,6 +316,34 @@ function timeKey(value) {
   ];
 }
 
+// A struct's integer field as a plain number, the way the ordering keys above read one.
+function fieldNumber(struct, name) {
+  return Number(structField(struct, name).value);
```
[the remark this answers is not in the document]
Comment (bartblast) [reply assets/js/elixir/hologram/entity.mjs:321]: Valid, fixed on both tiers - but the mechanism named here is not what happens, and that changed the fix.

The server does not reject a fractional component, it **raises**: `Calendar.ISO.valid_date?/3` and `valid_time?/4` are defined for integers only, so `%Date{month: 6.5}` came out of `Entity.Validator.validate/2` as a `FunctionClauseError` rather than as `{:error, _}`. That was a regression this PR introduced - the clause it replaced was `is_struct(value, Date)`, which answered `true`. So the server half was the more serious one, and it is fixed by giving each clause head `Calendar.ISO`'s own guard, letting a non-integer fall through to the `false` catch-all already there.

On the client the check reads the **boxed term** (`Type.isInteger`) rather than the converted number, which is the one place I did not follow the suggestion. `Number()` erases the distinction: a `%Date{month: 6.0}` satisfies `Number.isInteger` while the server's `is_integer/1` refuses it, so a `Number.isInteger` check would leave the tiers disagreeing about exactly the value the fix exists for. A mutation swapping the term check for the number check fails three of the new tests.

Two scope notes: nil components are not reachable (`Date`, `Time` and `DateTime` all enforce their component keys, so `%Date{}` will not build), and a microsecond that is not a pair is now refused too - the server answers that by its clause head failing to match.

Tests added per tier and per type, covering a float, a whole-valued float, a string, an atom, both halves of the microsecond pair, and a non-tuple microsecond.

File: test/javascript/elixir/hologram/entity_test.mjs [hunk 2]
```diff
@@ -879,6 +952,47 @@ describe("Elixir_Hologram_Entity", () => {
...
+
+      assert.deepEqual(item({opens_at: timeAtPrecision(7)}), typeViolation);
+      assert.deepEqual(item({opens_at: timeAtPrecision(6)}), Type.atom("ok"));
+    });
+
+    // The last microsecond of the day is a real time, so it reaches the CONSTRAINT pass and is
+    // refused by the declared maximum - which a type violation would have suppressed.
+    it("accepts the clock's last instant, which the declared maximum then refuses", () => {
+      assert.deepEqual(
+        item({opens_at: boxedTime(23, 59, 999999)}),
```
[the remark this answers is not in the document]
Comment (bartblast) [reply test/javascript/elixir/hologram/entity_test.mjs]: Correct, and it was hiding more than a mislabelled case. `boxedTime` hardcoded `second: 0`, so that call meant `23:59:00.999999` and the test's name ("the clock's last instant") was not true of the value it passed.

Chasing it showed the second field's range was untested on this path on BOTH tiers - widening `second <= 59` to `second <= 99` in `calendar.mjs` left `entity_test.mjs` fully green, and no Elixir test refused a second of 60 either. Only `model_test.mjs` caught it, through the box path. So the fix is a refusal case per tier as well as the helper change.

`boxedTime` takes `second` in clock order (`hour, minute, second, microsecond`) rather than appended last - the three call sites that passed a microsecond third are updated. A misleading argument order is what caused this, so preserving it seemed the wrong thing to optimise for.

File: lib/hologram/entity/validator.ex [hunk 3]
```diff
@@ -75,9 +75,41 @@ defmodule Hologram.Entity.Validator do
...
-  def attribute_value_valid?(value, :datetime, _opts), do: is_struct(value, DateTime)
+  # Date.new/3 will not build an impossible date, but a struct literal will, and so will a map
+  # update - both are ordinary Elixir. So the fields are asked of the struct's own calendar, which
+  # is the question Date.new/3 itself asks.
+  def attribute_value_valid?(
+        %Date{calendar: calendar, year: year, month: month, day: day},
+        :date,
+        _opts
+      ) do
+    calendar.valid_date?(year, month, day)
```
[the remark this answers is not in the document]
Comment (bartblast) [reply lib/hologram/entity/validator.ex:92]: Same defect as r3942512516, seen from the server side, and it is the half that mattered more: the clause heads now carry `Calendar.ISO`'s own `is_integer` guards, so a non-integer component falls through to the `false` catch-all and `validate/2` answers `{:error, [{name, {:type, type}}]}` where it raised. Regression cases per type are in, covering a float, a whole-valued float, a string, an atom and both halves of the microsecond pair.

The "calendar without the required callback" half is deliberately not guarded. A struct whose `calendar` names a module that is not a calendar raises `UndefinedFunctionError` from `Date.new/4`, `Date.to_string/1` and everything else Elixir has for it, so it is not a date in any sense the language recognises rather than a well-formed struct holding a bad value. Non-ISO calendars are out of scope for the data layer by design, and `calendar:` defaults, so reaching this takes a deliberate literal where a float month takes none. A `Code.ensure_loaded?` plus `function_exported?` on every validation would buy nothing for that price.

**Comment by bartblast** [comment]
@coderabbitai full review

File: assets/js/calendar.mjs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+// and Model's box helpers, which refuse a value the wire hands back. model.mjs is a base module
+// that entity.mjs imports, so the helpers cannot live in entity.mjs without inverting that.
+//
+// Split the way Elixir splits it - validDate against valid_date?/3, daysInMonth against
+// days_in_month/2, leapYear against leap_year?/1 - so a later divergence shows up in the half that
+// moved.
+export default class Calendar {
+  // valid_date?/3 is `is_month(month) and day in 1..days_in_month(year, month)`. The year is
+  // unconstrained: is_year/1 asks only that it is an integer.
+  static validDate(year, month, day) {
```
[the remark this answers is not in the document]
Comment (bartblast) [reply assets/js/calendar.mjs:26]: Accepted. Both helpers now refuse a fractional component through `Number.isInteger`, and the module gained its own test file - the check is unreachable through either caller, so nothing but a direct test could bind it, and `calendar.mjs` had none until now.

One limit is stated in the header rather than papered over: `Number.isInteger(6.0)` is true, because a JavaScript number cannot tell a whole-valued float from an integer. That half of Elixir's `is_integer/1` is refused where the boxed term is still in hand - the entity validator's `Type.isInteger` check from the earlier thread - and the new test file pins the limit explicitly so nobody reads this module as the whole guard.

The direct tests also closed a gap the review did not name: the century leap rule (1900 and 2100 not leap, 2000 leap) was pinned nowhere in the suite, since every earlier case used 2024 or 2026 and never reached the `% 100` / `% 400` half.
